nemo/collections/cv/models/visual_ctc_models.py

```diff
@@ -0,0 +1,694 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import copy
+import json
+import os
+import tempfile
+from math import ceil
+from typing import Dict, List, Optional, Union
+
+import torch
+from omegaconf import DictConfig, OmegaConf, open_dict
+from pytorch_lightning import Trainer
+from tqdm.auto import tqdm
+
+from nemo.collections.asr.data import audio_to_text_dataset
+from nemo.collections.asr.losses.ctc import CTCLoss
+from nemo.collections.asr.metrics.wer import WER, CTCDecoding, CTCDecodingConfig
+from nemo.collections.asr.models.asr_model import ASRModel, ExportableEncDecModel
+from nemo.collections.asr.parts.mixins import ASRModuleMixin, InterCTCMixin
+from nemo.collections.asr.parts.utils.audio_utils import ChannelSelectorType
+from nemo.collections.cv.data import video_to_text_dataset
+from nemo.core.classes.common import PretrainedModelInfo, typecheck
+from nemo.core.classes.mixins import AccessMixin
+from nemo.core.neural_types import LabelsType, LengthsType, LogprobsType, NeuralType, VideoSignal
+from nemo.utils import logging
+
+__all__ = ['VisualEncDecCTCModel']
+
+
+class VisualEncDecCTCModel(ASRModel, ExportableEncDecModel, ASRModuleMixin, InterCTCMixin):
+    """Base class for encoder decoder CTC-based models."""
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None):
+
+        # Get global rank and total number of GPU workers for IterableDataset partitioning, if applicable
+        # Global_rank and local_rank is set by LightningModule in Lightning 1.2.0
+        self.world_size = 1
+        if trainer is not None:
+            self.world_size = trainer.world_size
+
+        # Init
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        # Preprocessors
+        self.video_preprocessor = VisualEncDecCTCModel.from_config_dict(self._cfg.video_preprocessor)
+
+        # Augmentations
+        self.video_augmentation = VisualEncDecCTCModel.from_config_dict(self._cfg.video_augment)
+
+        # Front-end Networks
+        self.video_front_end = VisualEncDecCTCModel.from_config_dict(self._cfg.video_front_end)
```

Comment:
Side note: Whats the difference between front end and preprocessor? Maybe make a note of that

Reply:
the preprocessor apply video transforms like grayscale / resize / normalisation while the front-end is part of the network and is learned. The front-end is used to transform videos to temporal sequences for the encoder. This can be any network (Resnet+SpatialPooling, Linear Layer, ...)

So the order is:
- video_preprocessor
- video_augmentation
- video_front_end
- encoder